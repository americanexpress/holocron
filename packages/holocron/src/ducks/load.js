/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { Map as iMap, Set as iSet } from 'immutable';
import { getModule, getModuleMap } from '../moduleRegistry';
import {
  HOLOCRON_STORE_KEY,
  MODULES_STORE_KEY,
  REDUCER_KEY,
  REGISTER_MODULE_REDUCER,
  MODULE_LOADED,
  MODULE_LOAD_FAILED,
  MODULE_LOADING,
  MODULE_REDUCER_ADDED,
} from './constants';

const initialState = iMap({
  withReducers: iSet(),
  loaded: iSet(),
  failed: iMap(),
  loading: iMap(),
});

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case REGISTER_MODULE_REDUCER: {
      const { moduleName } = action;
      return state.update('withReducers', iSet(), (withReducers) => withReducers.add(moduleName));
    }
    case MODULE_LOADED: {
      const { moduleName } = action;
      return state.withMutations((holocronState) => {
        holocronState
          .update('loaded', iSet(), (loaded) => loaded.add(moduleName))
          .update('loading', iMap(), (loading) => loading.delete(moduleName));
      });
    }
    case MODULE_LOAD_FAILED: {
      const { moduleName, error } = action;
      return state.withMutations((holocronState) => {
        holocronState
          .update('failed', iMap(), (failed) => failed.set(moduleName, error))
          .update('loading', iMap(), (loading) => loading.delete(moduleName));
      });
    }
    case MODULE_LOADING: {
      const { moduleName, promise } = action;
      return state.update('loading', iMap(), (loading) => loading.set(moduleName, promise));
    }

    default:
      return state;
  }
}

function registerModuleReducer(moduleName) {
  return {
    type: REGISTER_MODULE_REDUCER,
    moduleName,
  };
}

function moduleLoaded(moduleName) {
  return {
    type: MODULE_LOADED,
    moduleName,
  };
}

function moduleLoadFailed(moduleName, error) {
  return {
    type: MODULE_LOAD_FAILED,
    moduleName,
    error,
  };
}

function moduleLoading(moduleName, promise) {
  return {
    type: MODULE_LOADING,
    moduleName,
    promise,
  };
}

export function isLoaded(moduleName) {
  return (state) => state.hasIn([HOLOCRON_STORE_KEY, 'loaded', moduleName]);
}

export function failedToLoad(moduleName) {
  return (state) => state.hasIn([HOLOCRON_STORE_KEY, 'failed', moduleName]);
}

export function getLoadError(moduleName) {
  return (state) => state.getIn([HOLOCRON_STORE_KEY, 'failed', moduleName]);
}

export function isLoading(moduleName) {
  return (state) => !!state.getIn([HOLOCRON_STORE_KEY, 'loading', moduleName]);
}

export function getLoadingPromise(moduleName) {
  return (state) => state.getIn([HOLOCRON_STORE_KEY, 'loading', moduleName]);
}

/* eslint-disable global-require */
export function loadModule(moduleName) {
  return (dispatch, getState, { modules, rebuildReducer }) => {
    const state = getState();
    const moduleData = getModuleMap().getIn([MODULES_STORE_KEY, moduleName]);

    let loadPromise;
    if (isLoaded(moduleName)(state)) {
      return Promise.resolve(getModule(moduleName, modules));
    }

    if (failedToLoad(moduleName)(state)) {
      return Promise.reject(getLoadError(moduleName)(state));
    }

    if (isLoading(moduleName)(state)) {
      return getLoadingPromise(moduleName)(state);
    }

    if (!moduleData) {
      const moduleLoadError = new Error(`Could not load Module ${moduleName} because it does not exist in the Module Version Map`);
      dispatch(moduleLoadFailed(moduleName, moduleLoadError));
      return Promise.reject(moduleLoadError);
    }

    if (modules) {
      const module = getModule(moduleName, modules);
      loadPromise = module ? Promise.resolve(module) : Promise.reject(new Error(`Module ${moduleName} was not preloaded on server`));
    } else {
      loadPromise = require('../loadModule.web.js').default(moduleName, moduleData);
    }

    dispatch(moduleLoading(moduleName, loadPromise));

    return loadPromise
      .then(
        (module) => {
          if (module[REDUCER_KEY]) {
            dispatch(registerModuleReducer(moduleName));
            rebuildReducer();
            dispatch({ type: MODULE_REDUCER_ADDED });
          }
          dispatch(moduleLoaded(moduleName));
          return module;
        },
        (error) => {
          console.warn(error);
          dispatch(moduleLoadFailed(moduleName, error));
          return Promise.reject(error);
        }
      );
  };
}
/* eslint-enable global-require */
