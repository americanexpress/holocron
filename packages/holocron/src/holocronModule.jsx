/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import hoistStatics from 'hoist-non-react-statics';

import {
  LOAD_KEY,
  REDUCER_KEY,
  MODULES_STORE_KEY,
  INIT_MODULE_STATE,
} from './ducks/constants';

// Execute deprecated load function and provide deprecation message
export function executeLoad({ dispatch, load, ...restProps } = { }) {
  if (load) {
    console.warn('The \'load\' function in holocron has been deprecated. Please use \'loadModuleData\' instead.');
    return dispatch(load(restProps));
  }
  return undefined;
}

// Dispatch loadModuleData inside a thunk if it exists
export function executeLoadModuleData(loadModuleData, WrappedComponent, props) {
  const { dispatch, ...restProps } = props;
  if (loadModuleData) {
    return dispatch(async (_, getState, { fetchClient }) => loadModuleData({
      store: { dispatch, getState },
      fetchClient,
      ownProps: restProps,
      module: WrappedComponent,
    }));
  }
  return undefined;
}

export function getModuleName(WrappedComponent, name) {
  return WrappedComponent && (WrappedComponent.displayName || WrappedComponent.name || name);
}

export function getModuleDisplayName(name) {
  return `HolocronModule(${name})`;
}

export async function executeLoadingFunctions({
  // Provide loadModuleData function
  loadModuleData,
  // Provide WrappedComponent for loadModuleData
  WrappedComponent,
  // Frozen props as of when called
  frozenProps,
  // Provide loadCount to limit state changes
  loadCount,
  // Provide componentName for error messages
  componentName,
  // Provide component instance for getting current state of component post async
  hocInstance,
}) {
  try {
    // Call deprecated load function if exists
    await executeLoad(frozenProps);
    // Call loadModuleData if it exists
    await executeLoadModuleData(loadModuleData, WrappedComponent, frozenProps);
    // Modify state only when mounted and current loadCount is less or equal than previous loadCount
    if (hocInstance.mounted && hocInstance.state.loadCount <= loadCount) {
      hocInstance.setState({ status: 'loaded' });
    }
  } catch (error) {
    console.error(`Error while attempting to call 'load' or 'loadModuleData' inside Holocron module ${componentName}.`, error);
    if (hocInstance.mounted) {
      hocInstance.setState({ status: 'error' });
    }
  }
}

export default function holocronModule({
  name,
  reducer,
  load, // TODO remove in next major version
  shouldModuleReload,
  loadModuleData,
  mergeProps,
  options = {},
} = {}) {
  return function wrapWithHolocron(WrappedComponent) {
    class HolocronModuleWrapper extends React.Component {
      constructor(props) {
        super(props);
        this.state = {
          loadCount: 0,
          status: 'loading',
        };
        this.mounted = false;
      }

      componentDidMount() {
        this.mounted = true;
        // eslint-disable-next-line no-underscore-dangle
        if ((loadModuleData || load) && !global.__INITIAL_STATE__) {
          this.initiateLoad(0, this.props);
        }
      }

      // ignoring to support deprecated componentWillReceiveProps.
      // This needs to be removed to support React17
      // eslint-disable-next-line camelcase
      UNSAFE_componentWillReceiveProps(nextProps) {
        const { loadCount } = this.state;
        const { props } = this;
        if (shouldModuleReload && shouldModuleReload(props, nextProps)) {
          const newLoadCount = loadCount + 1;
          this.setState({ status: 'loading', loadCount: newLoadCount });
          this.initiateLoad(newLoadCount, nextProps);
        }
      }

      // Ignoring as internal state cannot be checked when unmounted.
      /* istanbul ignore next */
      componentWillUnmount() {
        this.mounted = false;
      }

      initiateLoad(loadCount, frozenProps) {
        return executeLoadingFunctions({
          loadModuleData,
          WrappedComponent,
          frozenProps,
          loadCount,
          hocName: getModuleName(WrappedComponent, name),
          hocInstance: this,
        });
      }

      render() {
        const { status } = this.state;
        // eslint-disable-next-line react/jsx-props-no-spreading
        return <WrappedComponent {...this.props} moduleLoadStatus={status} />;
      }
    }

    HolocronModuleWrapper.propTypes = {
      load: PropTypes.func,
      dispatch: PropTypes.func.isRequired,
    };

    HolocronModuleWrapper.defaultProps = {
      load: undefined,
    };

    HolocronModuleWrapper.displayName = getModuleDisplayName(getModuleName(WrappedComponent, name));

    if (load && (options.ssr || global.BROWSER)) {
      HolocronModuleWrapper[LOAD_KEY] = load;
    }

    let mapModuleStateToProps;

    if (reducer && !name) {
      console.warn(`The Holocron Config in '${getModuleDisplayName(getModuleName(WrappedComponent, name))}' requires a 'name' when passing a 'reducer'.\nThe 'reducer' will not be added to the Redux Store without a 'name'.`);
    }

    if (reducer && name) {
      HolocronModuleWrapper[REDUCER_KEY] = reducer;
      const getModuleState = createSelector(
        (state) => state.getIn(
          [MODULES_STORE_KEY, name],
          reducer(undefined, { type: INIT_MODULE_STATE })
        ),
        (moduleState) => moduleState.toJS()
      );

      mapModuleStateToProps = ((state) => {
        const moduleState = getModuleState(state);
        return { moduleState };
      });
    }

    const mapDispatchToProps = (dispatch) => {
      if (load) {
        return {
          load,
          dispatch,
        };
      }
      return { dispatch };
    };

    hoistStatics(HolocronModuleWrapper, WrappedComponent);

    const DerivedComponent = connect(
      mapModuleStateToProps,
      mapDispatchToProps,
      mergeProps
    )(HolocronModuleWrapper);

    hoistStatics(DerivedComponent, HolocronModuleWrapper);

    return DerivedComponent;
  };
}
