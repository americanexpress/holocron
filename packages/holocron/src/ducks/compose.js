/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { loadModule } from './load';
import { LOAD_KEY } from './constants';

// This duck does not have a reducer
export default null;

function selectLoadModuleData(module) {
  // TODO remove module.loadModuleData in next major version
  if (module && module.loadModuleData) {
    return module.loadModuleData;
  }
  if (module && module.holocron && module.holocron.loadModuleData) {
    return module.holocron.loadModuleData;
  }
  return undefined;
}

export function composeModules(moduleConfigs) {
  return (dispatch, getState, { fetchClient } = {}) => {
    const modulePromises = moduleConfigs.map((config) => {
      const { name } = config;
      return dispatch(loadModule(name))
        .then((module) => {
          if (module[LOAD_KEY]) {
            return dispatch(module[LOAD_KEY](config.props));
          }

          const loadModuleData = selectLoadModuleData(module);

          if (loadModuleData) {
            return loadModuleData({
              store: { dispatch, getState },
              module,
              ownProps: config.props,
              fetchClient,
            });
          }

          return Promise.resolve();
        })
        .catch((error) => error);
    });
    return Promise.all(modulePromises);
  };
}
