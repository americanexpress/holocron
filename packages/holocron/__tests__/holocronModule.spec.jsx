/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/prop-types, import/no-extraneous-dependencies  */
import React from 'react';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { combineReducers as immutableCombineReducers } from 'redux-immutable';
import { Provider, connect } from 'react-redux';
import { fromJS } from 'immutable';
import renderer from 'react-test-renderer';
import _ from 'lodash';

import holocronModule, {
  executeLoad,
  executeLoadModuleData,
  executeLoadingFunctions,
} from '../src/holocronModule';
import { REDUCER_KEY, LOAD_KEY } from '../src/ducks/constants';

const sleep = (ms) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});

const warn = jest.spyOn(console, 'warn');
const error = jest.spyOn(console, 'error');

describe('holocronModule', () => {
  let fakeLoadModuleData;
  let FakeComponent;
  let fakeDispatch;
  let fakeGetState;
  let fakeFetchClient;
  let fakeProps;
  let fakeSetState;
  let fakeInstance;
  let fakeLoad;
  beforeEach(() => {
    jest.resetAllMocks();
    FakeComponent = {
      displayName: 'FakeComponent',
    };
    fakeLoadModuleData = jest.fn(async () => undefined);
    fakeFetchClient = jest.fn();
    fakeGetState = jest.fn();
    fakeDispatch = jest.fn(
      (func) => func(fakeDispatch, fakeGetState, { fetchClient: fakeFetchClient })
    );
    fakeSetState = jest.fn();
    fakeProps = {
      dispatch: fakeDispatch,
      load: fakeLoad,
    };
    fakeInstance = {
      setState: fakeSetState,
      mounted: true,
      state: {
        loadCount: 0,
      },
    };
  });
  afterEach(() => {
    global.BROWSER = false;
  });

  describe('executeLoad', () => {
    it('should return undefined with no args', () => {
      expect(executeLoad()).toEqual(undefined);
    });

    it('should call load with props', () => {
      fakeProps = {
        dispatch: jest.fn(),
        load: jest.fn((arg) => arg),
        myFakeProp: true,
      };
      executeLoad(fakeProps);
      expect(warn).toHaveBeenCalled();
      expect(fakeProps.load).toHaveBeenCalledWith({ myFakeProp: true });
    });
  });

  describe('executeLoadModuleData', () => {
    it('should call loadModuleData with correct args', async () => {
      expect.assertions(2);
      await executeLoadModuleData(
        fakeLoadModuleData,
        FakeComponent,
        { dispatch: fakeDispatch }
      );
      expect(fakeLoadModuleData.mock.calls).toMatchSnapshot();
      expect(fakeProps.dispatch).toHaveBeenCalled();
    });
  });

  describe('executeLoadingFunctions', () => {
    it('should call setState with success', async () => {
      expect.assertions(1);
      await executeLoadingFunctions({
        loadModuleData: fakeLoadModuleData,
        WrappedComponent: FakeComponent,
        frozenProps: fakeProps,
        loadCount: 0,
        componentName: 'FakeComponent',
        hocInstance: fakeInstance,
      });
      expect(fakeSetState).toHaveBeenCalledWith({ status: 'loaded' });
    });

    it('should not call setState if over loadCount', async () => {
      expect.assertions(1);
      await executeLoadingFunctions({
        loadModuleData: fakeLoadModuleData,
        WrappedComponent: FakeComponent,
        frozenProps: fakeProps,
        loadCount: 0,
        componentName: 'FakeComponent',
        hocInstance: {
          ...fakeInstance,
          mounted: false,
          state: {
            loadCount: 9999,
          },
        },
      });
      expect(fakeSetState).not.toHaveBeenCalled();
    });

    it('should call setState with error on failure', async () => {
      expect.assertions(1);
      fakeLoadModuleData = () => {
        throw new Error('Failed');
      };
      await executeLoadingFunctions({
        loadModuleData: fakeLoadModuleData,
        WrappedComponent: FakeComponent,
        frozenProps: fakeProps,
        loadCount: 0,
        componentName: 'FakeComponent',
        hocInstance: fakeInstance,
      });
      expect(fakeSetState).toHaveBeenCalledWith({ status: 'error' });
    });

    it('should not call setState if unmounted on failure', async () => {
      expect.assertions(2);
      fakeLoadModuleData = () => {
        throw new Error('Failed');
      };
      await executeLoadingFunctions({
        loadModuleData: fakeLoadModuleData,
        WrappedComponent: FakeComponent,
        frozenProps: fakeProps,
        loadCount: 0,
        componentName: 'FakeComponent',
        hocInstance: {
          ...fakeInstance,
          mounted: false,
        },
      });
      expect(error).toHaveBeenCalled();
      expect(fakeSetState).not.toHaveBeenCalled();
    });
  });

  it('should wrap module with no arguments', () => {
    const MyModuleComponent = holocronModule()(() => <div>Mock Module</div>);
    const mockStore = createStore((state) => state, fromJS({ modules: { 'mock-module': { key: 'value' } } }));
    const tree = renderer.create(<MyModuleComponent store={mockStore} />);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should provide the module state as a plain JS prop if a reducer is provided', () => {
    const reducer = (state) => state;
    const Module = holocronModule({
      name: 'mock-module',
      reducer,
    })(({ moduleState }) => <div>{moduleState.key}</div>);
    const mockStore = createStore((state) => state, fromJS({ modules: { 'mock-module': { key: 'value' } } }));
    const component = renderer.create(<Module store={mockStore} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not rerender if module state has not changed', () => {
    const renderSpy = jest.fn();
    const moduleReducer = (state) => state || {};
    const InnerModule = () => { renderSpy(); return <div>Mock Module</div>; };
    const Module = holocronModule({
      name: 'mock-module',
      reducer: moduleReducer,
    })(InnerModule);
    const appReducer = (state, action) => (action.type === 'MOCK_ACTION_TYPE' ? action.newState : state || {});
    const reducer = immutableCombineReducers({
      app: appReducer,
      modules: immutableCombineReducers({
        'mock-module': moduleReducer,
      }),
    });
    const mockStore = createStore(
      reducer,
      fromJS({ app: { someParam: 'initial' }, modules: { 'mock-module': { key: 'value' } } })
    );
    renderer.create(<Module store={mockStore} />);
    mockStore.dispatch({ type: 'MOCK_ACTION_TYPE', newState: { someParam: 'new' } });
    expect(renderSpy).toHaveBeenCalledTimes(1);
  });

  it('should dispatch the module\'s load action on componentDidMount with no preloaded state', () => {
    global.INITIAL_STATE = undefined;
    const load = jest.fn(() => () => Promise.resolve());
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(() => <div>Mock Module</div>);
    const mockStore = createStore((state) => state, applyMiddleware(thunk));
    const props = { a: 'b', x: { y: 'z' } };
    renderer.create(
      <Provider store={mockStore}>
        <Module {...props} />
      </Provider>
    );
    // couldn't use toHaveBeenCalledWith because mapDispatchToProps is used
    const calledProps = load.mock.calls[0][0];
    const calledPropsWithoutFunctions = _.pickBy(calledProps, (p) => typeof p !== 'function');
    expect(calledPropsWithoutFunctions).toEqual(props);
    expect(load).toHaveBeenCalledTimes(1);
  });

  it('should not dispatch the module\'s load action on client takeover of a server render', () => {
    global.INITIAL_STATE = { some: 'state' };
    const load = jest.fn(() => () => Promise.resolve());
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(() => <div>Mock Module</div>);
    const mockStore = createStore((state) => state, applyMiddleware(thunk));
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    expect(load).not.toHaveBeenCalled();
  });

  it('should not dispatch the module\'s load action on the server render', () => {
    const load = jest.fn();
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(({ moduleLoadStatus }) => <div>Mock Module - {moduleLoadStatus}</div>);
    const mockStore = createStore((state) => state);
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    expect(load).not.toHaveBeenCalled();
  });

  it('should dispatch the module\'s load action if it receives new props that pass shouldModuleReload', () => {
    const load = jest.fn(() => () => Promise.resolve());
    const Module = connect((state) => state)(holocronModule({
      name: 'mock-module',
      load,
      shouldModuleReload: (currProps, nextProps) => currProps.someParam !== nextProps.someParam,
    })(() => <div>Mock Module</div>));
    const mockStore = createStore(
      (state, action) => (action.type === 'MOCK_ACTION_TYPE' ? action.newState : state),
      { someParam: 'initial' },
      applyMiddleware(thunk)
    );
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    mockStore.dispatch({ type: 'MOCK_ACTION_TYPE', newState: { someParam: 'new' } });
    // couldn't use toHaveBeenCalledWith because mapDispatchToProps is used
    const calledProps = load.mock.calls[0][0];
    const calledPropsWithoutFunctions = _.pickBy(calledProps, (p) => typeof p !== 'function');
    expect(calledPropsWithoutFunctions).toEqual({ someParam: 'new' });
    expect(load).toHaveBeenCalledTimes(1);
  });

  it('should not dispatch the module\'s load action if it receives new props that do not pass shouldModuleReload', () => {
    const load = jest.fn(() => ({ type: 'LOAD' }));
    const Module = connect((state) => state)(holocronModule({
      name: 'mock-module',
      load,
      shouldModuleReload: (currProps, nextProps) => currProps.someParam !== nextProps.someParam,
    })(() => <div>Mock Module</div>));
    const mockStore = createStore((state, action) => (action.type === 'MOCK_ACTION_TYPE' ? action.newState : state), { someParam: 'initial' });
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    const { dispatch } = mockStore;
    dispatch({ type: 'MOCK_ACTION_TYPE', newState: { someParam: 'initial', differentParam: 'new' } });
    expect(load).not.toHaveBeenCalled();
  });

  it('should not dispatch the module\'s load action if no shouldModuleReload function is provided', () => {
    const load = jest.fn(() => ({ type: 'LOAD' }));
    const Module = connect((state) => state)(holocronModule({
      name: 'mock-module',
      load,
    })(() => <div>Mock Module</div>));
    const mockStore = createStore((state, action) => (action.type === 'MOCK_ACTION_TYPE' ? action.newState : state), { someParam: 'initial' });
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    const { dispatch } = mockStore;
    dispatch({ type: 'MOCK_ACTION_TYPE', newState: { someParam: 'new' } });
    expect(load).not.toHaveBeenCalled();
  });

  // TODO: use enzyme to assert correct props, need to update version of jest first
  it('should pass the moduleLoadStatus prop as loading when loading', () => {
    global.INITIAL_STATE = undefined;
    const loadPromise = Promise.resolve();
    const load = jest.fn(() => () => loadPromise);
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(({ moduleLoadStatus }) => <div>Mock Module - {moduleLoadStatus}</div>);
    const mockStore = createStore((state) => state, applyMiddleware(thunk));
    const tree = renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  // TODO: use enzyme to assert correct props, need to update version of jest first
  it('should pass the moduleLoadStatus prop as loaded when loaded', async () => {
    global.INITIAL_STATE = undefined;
    const loadPromise = Promise.resolve();
    const load = jest.fn(() => () => loadPromise);
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(({ moduleLoadStatus }) => <div>Mock Module - {moduleLoadStatus}</div>);
    const mockStore = createStore(
      (state) => state, applyMiddleware(thunk.withExtraArgument({ fetchClient: jest.fn() }))
    );
    const tree = renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );

    // Wait one cycle of the event loop since we can't retrieve a promise from initiateLoad
    await sleep(1);

    return loadPromise
      .then(() => {
        expect(tree.toJSON()).toMatchSnapshot();
      });
  });

  // TODO: use enzyme to assert correct props, need to update version of jest first
  it('should pass the moduleLoadStatus prop as error when it failed to load', () => {
    global.INITIAL_STATE = undefined;
    const loadPromise = Promise.reject();
    const load = jest.fn(() => () => loadPromise);
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(({ moduleLoadStatus }) => <div>Mock Module - {moduleLoadStatus}</div>);
    const mockStore = createStore((state) => state, applyMiddleware(thunk));
    const tree = renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );

    return loadPromise
      .catch(() => {
        expect(tree.toJSON()).toMatchSnapshot();
      });
  });

  it('should not try to setState if it is not mounted', () => {
    // TODO: actually test this when we update jest and start using enzyme and we can unmount
  });

  it('should gracefully handle load not returning a Promise', () => {
    global.INITIAL_STATE = undefined;
    const load = jest.fn(() => () => 'not a promise');
    const Module = holocronModule({
      name: 'mock-module',
      load,
    })(({ moduleLoadStatus }) => <div>Mock Module - {moduleLoadStatus}</div>);
    const mockStore = createStore((state) => state, applyMiddleware(thunk));
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    // Don't need to expect anything, it passes if no errors are thrown
  });

  it('should have a helpful display name', () => {
    const name = 'mock-module';
    const SomeComponent = () => <div />;
    SomeComponent.displayName = 'DisplayName';
    function FuncName() { return <div />; }
    expect(holocronModule({ name })(SomeComponent).displayName).toBe('Connect(HolocronModule(DisplayName))');
    expect(holocronModule({ name })(FuncName).displayName).toBe('Connect(HolocronModule(FuncName))');
    expect(holocronModule({ name })(() => <div />).displayName).toBe('Connect(HolocronModule(mock-module))');
  });

  it('should attach the load function as a static if it is on the browser', () => {
    global.BROWSER = true;
    const InnerModule = () => <div>Mock Module</div>;
    const load = () => null;

    const BrowserModule = holocronModule({
      name: 'mock-module',
      load,
    })(InnerModule);

    expect(BrowserModule[LOAD_KEY]).toBe(load);
  });

  it('should attach the load function as a static on the server if the ssr option is set', () => {
    const InnerModule = () => <div>Mock Module</div>;
    const load = () => null;

    const ModuleWithSSROption = holocronModule({
      name: 'mock-module',
      load,
      options: { ssr: true },
    })(InnerModule);

    expect(ModuleWithSSROption[LOAD_KEY]).toBe(load);
  });

  it('should not attach the load function as a static on the server if ssr option is not set', () => {
    const InnerModule = () => <div>Mock Module</div>;
    const load = () => null;

    const ModuleWithoutSSROption = holocronModule({
      name: 'mock-module',
      load,
    })(InnerModule);

    expect(ModuleWithoutSSROption[LOAD_KEY]).toBeUndefined();
  });

  it('should add reducer as a static', () => {
    const InnerModule = () => <div>Mock Module</div>;
    const reducer = (state) => state;
    const Module = holocronModule({
      name: 'mock-module',
      reducer,
    })(InnerModule);
    expect(Module[REDUCER_KEY]).toBe(reducer);
  });

  it('should hoist non react statics', () => {
    const InnerModule = () => <div>Mock Module</div>;
    InnerModule.someStatic = 'static property';
    const Module = holocronModule({
      name: 'mock-module',
    })(InnerModule);
    expect(Module.someStatic).toBe('static property');
  });

  it('should pass mergeProps to connect if it is specified', () => {
    const reducer = () => fromJS({ x: 3 });
    const mergeProps = (stateProps, dispatchProps, ownProps) => {
      const xy = stateProps.moduleState.x * ownProps.y;
      return {
        ...stateProps,
        ...dispatchProps,
        ...ownProps,
        xy,
      };
    };
    const Module = holocronModule({
      name: 'mock-module',
      reducer,
      mergeProps,
    })(({ moduleState: { x }, y, xy }) => <div>{x} * {y} = {xy}</div>);
    const mockStore = createStore((state) => state, fromJS({ modules: { 'mock-module': { x: 3 } } }));
    const component = renderer.create(
      <Provider store={mockStore}>
        <Module y={4} />
      </Provider>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should gracefully handle module state missing from store', () => {
    const reducer = () => fromJS({ x: 2 });
    const Module = holocronModule({
      name: 'mock-module',
      reducer,
    })(({ moduleState }) => <div>Mock Module {moduleState.x}</div>);
    const mockStore = createStore((state) => state, fromJS({ modules: {} }));
    const render = () => renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    expect(render).not.toThrowError();
    expect(render().toJSON()).toMatchSnapshot();
  });
  it('should warn if a reducer is set but no name', () => {
    const reducer = () => {};
    const Module = holocronModule({
      reducer,
    })(() => <div>Mock Module</div>);
    const mockStore = createStore((state) => state, fromJS({}));
    renderer.create(
      <Provider store={mockStore}>
        <Module />
      </Provider>
    );
    expect(warn.mock.calls).toMatchSnapshot();
  });
});
