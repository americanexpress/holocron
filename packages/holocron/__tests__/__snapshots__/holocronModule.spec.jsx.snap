// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`holocronModule executeLoadModuleData should call loadModuleData with correct args 1`] = `
Array [
  Array [
    Object {
      "fetchClient": [MockFunction],
      "module": Object {
        "displayName": "FakeComponent",
      },
      "ownProps": Object {},
      "store": Object {
        "dispatch": [MockFunction] {
          "calls": Array [
            Array [
              [Function],
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Promise {},
            },
          ],
        },
        "getState": [MockFunction],
      },
    },
  ],
]
`;

exports[`holocronModule should gracefully handle module state missing from store 1`] = `
<div>
  Mock Module 
  2
</div>
`;

exports[`holocronModule should pass mergeProps to connect if it is specified 1`] = `
<div>
  3
   * 
  4
   = 
  12
</div>
`;

exports[`holocronModule should pass the moduleLoadStatus prop as error when it failed to load 1`] = `
<div>
  Mock Module - 
  error
</div>
`;

exports[`holocronModule should pass the moduleLoadStatus prop as loaded when loaded 1`] = `
<div>
  Mock Module - 
  loaded
</div>
`;

exports[`holocronModule should pass the moduleLoadStatus prop as loading when loading 1`] = `
<div>
  Mock Module - 
  loading
</div>
`;

exports[`holocronModule should provide the module state as a plain JS prop if a reducer is provided 1`] = `
<div>
  value
</div>
`;

exports[`holocronModule should warn if a reducer is set but no name 1`] = `
Array [
  Array [
    "The Holocron Config in 'HolocronModule(undefined)' requires a 'name' when passing a 'reducer'.
The 'reducer' will not be added to the Redux Store without a 'name'.",
  ],
]
`;

exports[`holocronModule should wrap module with no arguments 1`] = `
<div>
  Mock Module
</div>
`;
